### PR TITLE
[WIP] Bar Label

### PR DESCRIPTION
**Pull request by kanitw**
@willium let’s create a feature branch “1.x-label” so that we can merge label related stuff into this branch since we’re not sure if all labels will be done by the 1.0 release.

### CODE REVIEW

File: src/compiler/marks.ts [hunk 1]
```diff
@@ -357,8 +357,53 @@ export namespace bar {
...
+    // - color support
+    // - container height/stretching consideration
+    // - consider defaults for offset
+
+    var l: any = {};
+    const orient = model.marksConfig('orient');
+
+    l.align = { value: (orient !== 'horizontal') ? "center" : "left" };
+
+    l.text = { field: model.has(LABEL) ? model.field(LABEL) : ((orient !== 'horizontal') ? model.field(Y) : model.field(X)) };
```
Comment (kanitw) [inline src/compiler/marks.ts]: if there is not `label` channel, this function would not be evaluated at all.   

See [this line](https://github.com/vega/vega-lite/blob/master/src/compiler/marks.ts#L104)

Comment (willium) [inline src/compiler/marks.ts]: right, I meant to try to pick up on whether or not the VL explicitly assigns a __field__ (see below) to the label, but you're right the way I wrote it is redundant. 

     "encoding": {
                "x": {"type": "quantitative", "field": "amount"},
                "y": {"type": "ordinal", "field": "category"},
                "label": {
                  "field": "amount",
                  "type": "quantitative"
                }
            }

**Comment by willium** [comment]
I'm curious if you could point me to where I could pick up the model's colors (bg, where steelblue comes from, axes). I figure the labels are better default colored to the axes, though I imagine there might be an algorithm for finding the best color (you probably already have it implemented somewhere)

There are likely some bigger conversations to be had or that you've already had, but It'd be cool to work on some of the other labels and variations. 
Some thoughts I had:
- inset (negatively colored) labels could be cool, but would cause problems with stacked bars (unless there was a label inset for each stacked bar, but then you'd have a width/height limitation
- the fix you mentioned in your (comment about intelligently scaling the bounds)[https://github.com/vega/vega-lite/issues/64#issuecomment-158663494] probably wouldn't work for relative bar charts once they're maxed out to 100%
- it'd be cool to have a short hand for specific points in scatterplots: outliers, critical points and summary numbers, or intersections with line charts, but all these labels would probably look different and be more complicated than just a floating annotation

File: src/compiler/marks.ts (hunk 1 above)
Comment (kanitw) [inline src/compiler/marks.ts]: If you look at `model.has()` in `Model.ts` -- `has` actually check existence of the field.   So we can be sure about that. 

(I start wondering if we should rename `has` to `hasField`..)

Comment (willium) [inline src/compiler/marks.ts]: ah you're right. I should've looked more closely. I wonder if there's a case for wanting to apply a label without being explicit with field. In the case of a bar chart with a quantitative variable and a ordinal/categorical variable, I can't imagine the label would be of anything but the quantitative var. Either way, totally makes sense, I'll remove that conditional.

**Comment by kanitw** [comment]
> where I could pick up the model's colors (bg, where steelblue comes from, axes).

It's actually very simple.   If color channel is not mapped to a field, we draw value from `encoding.color.value`.  If `encoding.color.value` is not explicitly specified, we use the default value from [this line](https://github.com/vega/vega-lite/blob/master/src/schema/encoding.schema.ts#L74).  (`vl.schema.instantiate` uses `default` property in the schema to set default values for properties.) 


> - inset (negatively colored) labels could be cool, but would cause problems with stacked bars (unless there was a label inset for each stacked bar, but then you'd have a width/height limitation)

This is definitely a good thing to do in the future since there are many corner cases like you said. 
I think it's worth do labels for other mark types first before introducing advanced features for one mark type (bar).   

Other mark types will have complicated issues too, but maybe we can start with basic supports for all mark types before upgrading rules for each specific ones.  For example, scatterpoint's labels can lead to a lot of occlusion, but we might assume that the first version we are okay with occlusion.  In a later version, we can then introduce a Vega transform for hiding occluded labels.  

> the fix you mentioned in your [comment about intelligently scaling the bounds](https://github.com/vega/vega-lite/issues/64#issuecomment-158663494) probably wouldn't work for relative bar charts once they're maxed out to 100%

Good point.  In that case, we may just add offset and hope that Vega will support [padding at the group level soon](https://github.com/vega/vega/issues/472).   Once we have inset labels, we may be have relative bar charts use inset labels by default.  

> - it'd be cool to have a short hand for specific points in scatterplots: outliers, critical points and summary numbers, or intersections with line charts, but all these labels would probably look different and be more complicated than just a floating annotation

Yeah, these would be useful in the future, but is much more complicated to do now.  We have to implement Vega transforms for detecting outliers before we can highlight/visualize them.  

>  intersections with line charts
(I'm not sure what you mean "intersections with line charts" though)

**Comment by kanitw** [comment]
FYI, I just added more comment about axis’s color in the comment above.

**Comment by willium** [comment]
- [x] I fixed the conditional

Because the axis color is set by Vega, how should we handle the coloring of the label? I'm not sure if I can piggy back on some other logic (the grid perhaps?), or should write something more custom. Not sure what the desired behavior is either for labels.

I hardcoded in offsets of 6 because its the same as the grid offset, but wasn't sure if I should directly pull from that or not... Also, for the horizontal bars, the __l.dy__'s _mult_ and _band_ parameters don't seem to do anything. If I hardcode in half the grid offset as a y offset, it does the trick but -- I'm not sure we can be sure that the bar's widths are only 6, especially in such a hardcoded way. Is there something to _mult_ and _band_ not working that I should realize, or is that a bug? I'm not sure what numbers I can rely on in this case.

**Comment by kanitw** [comment]
> Because the axis color is set by Vega, how should we handle the coloring of the label?

This is marks label, so we shouldn't worry too much about axis label.   That said, I think having label in black by default is a reasonable option.  We can then provide a set of label configs ( `config.label.<property>` for customizing marks property of the label marks and other label options).  
And maybe add a special `config.label.sameColorAsMark` to make the label use the same color as mark (no matter it's from `encoding.color.value` or `encoding.color.field`)

I just send a PR to provide scaffolding for `config.label`.  

> I hardcoded in offsets of 6
Eventually we might add `config.label.offset` as part of `config.label`.  
we might try to add label for other mark type first -- just to see if they can share same set of config properties? 

> I hardcoded in offsets of 6 because its the same as the grid offset, but wasn't sure if I should directly pull from that or not.

I'm changing gridOffset to 0 and they are different things -- so definitely not! 

>  l.dy's mult and band parameters don't seem to do anything
Maybe @arvind  will know better about this.

File: src/compiler/marks.ts [hunk 2]
```diff
@@ -357,8 +357,48 @@ export namespace bar {
...
+      l.y.offset = 3;
+      l.dy = {
+        scale: model.scale(Y),
+        band: true,
+        mult: 0.5
+      }
+    }
+
+    // fill
+    extend(l, colorMixins(model));
```
Comment (willium) [inline src/compiler/marks.ts]: `applyColorAndOpacity(l, model)`

File: src/compiler/marks.ts [hunk 3]
```diff
@@ -357,8 +357,48 @@ export namespace bar {
...
 
   export function labels(model: Model) {
-    // TODO(#64): fill this method
-    return undefined;
+    // TODO(#64):
+    // - adjust for container bounds
+    // - consider defaults for offset and color
+
+    var l: any = {};
+    const orient = model.marksConfig('orient');
```
Comment (willium) [inline src/compiler/marks.ts]: model.config().mark.orient

**Comment by kanitw** [comment]
@willium 

A few notes from today’s discussion 
- VL’s bar use `xc` and `width` instead of  `x` and `x2` with offset.  
- We talked about reactive geometry, which means you can drive mark with another mark like shown in the [text group in the group bar example](http://vega.github.io/vega-editor/?mode=vega&spec=grouped_bar).    There is no doc for reactive geometry yet but the note in https://github.com/vega/vega/issues/497 should be useful.

**Pull request closed without merging**